### PR TITLE
Orca: refactor testing py directories by backends

### PR DESCRIPTION
**Pull request by hamham223**
### 1. Why the change?

Some spark backend testing pys are under `python/orca/test/bigdl/orca/learn/ray`, which brings ambiguity. This PR tries to refactor the directory `orca/learn` into 3 categories based on the backend
+ `bigdl`
+ `spark`
+ `ray`
+ `jep` ( also `bigdl` backend)

### 2. Summary of the change 

+ rename `orca/test/learn/spark/` into `orca/test/learn/bigdl`, all the testings are bigdl backend
+ move the location of two testing python scripts
    | file name | package requirements | original location | location in this PR| original testing shell | 
    | :-----------: | ------------ | ------------ | ------------ | --------- | 
    | `test_tf_spark_estimator.py` ➡️  **`test_tf2_estimator.py`**| tf2 | [`orca/learn/ray/tf/`](https://github.com/intel-analytics/BigDL/blob/main/python/orca/test/bigdl/orca/learn/ray/tf/test_tf_spark_estimator.py) | `orca/learn/spark/` | [`run-pytests-horovod-tf.sh`](https://github.com/intel-analytics/BigDL/blob/main/python/orca/dev/test/run-pytests-horovod-tf.sh) | 
    | `test_estimator_pyspark_backend.py` ➡️  **`test_pytorch_estimator.py`** | torch + pandas + tensorboard  |[`orca/learn/ray/pytorch/`](https://github.com/intel-analytics/BigDL/blob/main/python/orca/test/bigdl/orca/learn/ray/pytorch/test_estimator_pyspark_backend.py) | `orca/learn/spark/` | [`run-pytests-ray.sh`](https://github.com/intel-analytics/BigDL/blob/main/python/orca/dev/test/run-pytests-ray.sh#L44) |
+ `orca/dev/test`: change corresponding testing shell scripts
    -  **they are now both tested in `run-pytests-ray.sh`**
    - should both pass py37 and py38 tests
+ `orca/developer_guide.md`: update

```markdown
   3. Run tests to verify environment
   
   cd BigDL/python/orca
   pytest test/bigdl/orca/learn/spark/test_estimator_pyspark_backend.py
   ->
   pytest test/bigdl/orca/learn/spark/test_pytorch_estimator.py
   ```
+ `PR_validation.yml` and corresponding `action.yml`

### 3. How to test?
The following unit test must be completed ✅ 
- [ ] Orca-Pytest-Py37-Spark3-Tf1
- [ ] Orca-Pytest-Py38-Spark3
- [ ] Orca-Pytest-Ray-Py37-Spark3
- [ ] Orca-Pytest-Ray-Py38-Spark3

### CODE REVIEW

File: python/orca/test/bigdl/orca/learn/spark/test_tf_spark_estimator.py [hunk 1]
```diff
@@ -27,8 +27,6 @@
 from bigdl.orca import OrcaContext
 
 
-resource_path = os.path.join(
-    os.path.realpath(os.path.dirname(__file__)), "../../../../resources")
```
Comment (hamham223) [inline python/orca/test/bigdl/orca/learn/spark/test_tf_spark_estimator.py:31]: This variable is not used, thus removed.

File: python/orca/dev/test/run-pytests-spark.sh [hunk 2]
```diff
@@ -30,8 +30,8 @@ export PYSPARK_DRIVER_PYTHON=python
 python_version=$(python --version | awk '{print$2}')
 if [ $python_version == 3.7.10 ];then
   python -m pytest -v test/bigdl/orca/tfpark
-  python -m pytest -v test/bigdl/orca/learn/spark --ignore=test/bigdl/orca/learn/spark/test_estimator_openvino.py
-
+  python -m pytest -v test/bigdl/orca/learn/spark --ignore=test_tf_spark_estimator.py
```
Comment (hamham223) [inline python/orca/dev/test/run-pytests-spark.sh:33]: should be `test/bigdl/orca/learn/spark/test_tf_spark_estimator.py`

**Comment by hamham223** [comment]
This PR might have conflicts with other PRs, thus turn to draft

File: .github/workflows/PR_validation.yml [hunk 3]
```diff
@@ -313,7 +313,26 @@ jobs:
...
+      uses: ./.github/actions/maven-setup-action
+    - name: Setup env
+      uses: ./.github/actions/orca/setup-env/setup-orca-python-py37-spark3
+    - name: Run Test
+      uses: ./.github/actions/orca/orca-python-tf2-py37-spark3-action
+    - name: Remove Env
+      if: ${{ always() }}
+      uses: ./.github/actions/remove-env
+
+  Orca-Python-Ray-Py37-Spark3:
```
Comment (hamham223) [inline .github/workflows/PR_validation.yml:335]: WIP: The conflict is handled incorrectly..

Comment (hamham223) [reply .github/workflows/PR_validation.yml]: fixed

File: python/orca/dev/test/run-pytests-ray.sh [hunk 4]
```diff
@@ -49,23 +49,31 @@ python -m pytest -v test/bigdl/orca/learn/ray \
...
-exit $exit_status_4
+    exit $exit_status_4
 fi
 
 ray stop -f
+
+echo "Running orca learn tf2 spark backend tests"
+python -m pytest -v test/bigdl/orca/learn/spark/test_tf_spark_estimator.py
+exit_status_5=$?
+if [ $exit_status_5 -ne 0 ];
```
Comment (hamham223) [inline python/orca/dev/test/run-pytests-ray.sh:76]: Actually `if [ $exit_status_5 -ne 0 ];` isn't necessary? `set -ex` makes the scripts exit automatically on some errors. But it's beautiful to keep consistency! 😆

File: python/orca/dev/test/run-pytests-spark.sh [hunk 5]
```diff
@@ -37,12 +37,13 @@ if [ $python_version == 3.7.10 ];then
...
       exit $exit_status_1
   fi
-  python -m pytest -v test/bigdl/orca/learn/spark --ignore=test/bigdl/orca/learn/spark/test_estimator_openvino.py
+  python -m pytest -v test/bigdl/orca/learn/bigdl --ignore=test/bigdl/orca/learn/bigdl/test_estimator_openvino.py
   exit_status_2=$?
   if [ $exit_status_2 -ne 0 ];
   then
       exit $exit_status_2
   fi
+  python -m pytest -v test/bigdl/orca/learn/spark/test_estimator_pyspark_backend.py
```
Comment (hamham223) [inline python/orca/dev/test/run-pytests-spark.sh:46]: require tf1, thus only tested in python 3.7

File: python/orca/dev/test/run-pytests-ray.sh [hunk 6]
```diff
@@ -49,23 +49,31 @@ python -m pytest -v test/bigdl/orca/learn/ray \
...
 if [ $exit_status_2 -ne 0 ];
 then
-   exit $exit_status_2
+    exit $exit_status_2
 fi
 
 echo "Running orca learn tf2 ray backend tests"
 python -m pytest -v test/bigdl/orca/learn/ray/tf/test_tf2estimator_ray_backend.py
 exit_status_3=$?
 if [ $exit_status_3 -ne 0 ];
```
Comment (hkvision) [inline python/orca/dev/test/run-pytests-ray.sh:58]: Can we refactor this?

File: python/orca/dev/test/run-pytests-ray.sh [hunk 7]
```diff
@@ -49,23 +49,31 @@ python -m pytest -v test/bigdl/orca/learn/ray \
 exit_status_2=$?
 if [ $exit_status_2 -ne 0 ];
 then
-   exit $exit_status_2
+    exit $exit_status_2
 fi
 
 echo "Running orca learn tf2 ray backend tests"
```
Comment (hkvision) [inline python/orca/dev/test/run-pytests-ray.sh:55]: Also need to modify Line 44?

File: python/orca/test/bigdl/orca/learn/spark/conftest.py [hunk 8]
```diff
@@ -13,13 +13,28 @@
...
 
 
 @pytest.fixture(autouse=True, scope='package')
 def orca_context_fixture():
-    from bigdl.orca import init_orca_context
-    sc = init_orca_context(cores=4, spark_log_level="INFO")
-    yield sc
-    sc.stop()
+    from bigdl.orca import init_orca_context, stop_orca_context
+    conf = {"spark.python.worker.reuse": "false"}
```
Comment (hkvision) [inline python/orca/test/bigdl/orca/learn/spark/conftest.py:24]: we can remove this and run once to confirm the error?

Comment (hamham223) [reply python/orca/test/bigdl/orca/learn/spark/conftest.py:24]: ok, let's try and cry 😢

File: python/orca/dev/test/run-pytests-ray.sh (hunk 7 above)
Comment (hkvision) [reply python/orca/dev/test/run-pytests-ray.sh:55]: Can refactor this in the next PR.

**Review by hkvision — APPROVED** [review]

**Comment by hkvision** [comment]
One failure is irrelevant. Merge it first.

**Pull request merged**